File: backtracking/permutations.js
function fact(n) {
  let res = 1;

  for (let k = 1; k <= n; k++) {
    res *= k;
  }

  return res;
}

/**
 * Given an array, return an array containing every possible permutation
 * of the elements in the input array.
 *
 * If the input array has length N, what will the length of the returned
 * array be?
 */
function permutationsSwap(array, index = 0, results = []) {
  if (index === array.length) {
    results.push([...array]);

    return results;
  }

  // The "trick" here is that because we build permutations prefix-first, i.e.,
  // from left to right, we can use part of the original array to store
  // the permutation we've generated so far, and everything to the right of that
  // to store the remaining elements. We swap elements between the two parts,
  // which is a constant-time operation.
  for (let i = index; i < array.length; i++) {
    [array[index], array[i]] = [array[i], array[index]];

    permutations(array, index + 1, results);

    [array[i], array[index]] = [array[index], array[i]];
  }

  return results;
}


function permutations(array) {
  // If we call permutations(foo), we expect it to return an array
  // that contains foo (and potentially other arrays). That means
  // permutations([]) should return [[]] and not just []
  if (array.length === 0) {
    return [[]];
  }

  const results = [];

  for (let i = 0; i < array.length; i++) {
    const item = array[i];
    const remainingItems = array.slice(0, i).concat(array.slice(i + 1));

    const remainingPerms = permutations(remainingItems);

    for (let perm of remainingPerms) {
      results.push([item, ...perm]);
    }
  }

  return results;
}

function permutationsBad(array, results = []) {
  if (array.length === 0) {
    console.log(results);
    return;
  }

  for (let i = 0; i < array.length; i++) {
    let item = array[i];
    let remainingItems = array.slice(0, i).concat(array.slice(i + 1));

    permutationsBad(remainingItems, results.concat([item]));
  }
}

// Test

let array = Array.from('ABCD');
let allPermutations = permutations(array);

console.log(array);
console.log('');
console.log(allPermutations);

console.log('');
console.log('Expected # permutations: ', fact(array.length));
console.log('Actual # permutations:   ', allPermutations.length);

File: backtracking/permutations.py
def permutations_print(array, results = ''):
    if len(array) == 0:
        print(results)
        return

    for i, item in enumerate(array):
        without_item = array[0:i] + array[i+1:]
        permutations_print(without_item, results + item)

def permutations_collect(array):
    if (len(array) == 0):
        return [[]]

    results = []
    for i in range(len(array)):
        item = array[i]
        remaining_items = array[0:i] + array[i+1:]
        remaining_perms = permutations_collect(remaining_items)

        for perm in remaining_perms:
            results.append([item] + perm)

    return results


def swap(array, i, j):
    [array[i], array[j]] = [array[j], array[i]]
    return array

def permutations_swap_print(array, idx = 0):
    if (len(array) == idx):
        print(array)
        return

    for i in range(idx, len(array)):
        swap(array, i, idx)
        permutations_swap_print(array, idx + 1)
        swap(array, i, idx)

example_array = list('abc')

from pprint import pprint

# permutations_print(example_array, '')
# permutations_swap_print(example_array)
# pprint(permutations_collect(example_array), width=40)
